### PR TITLE
Class for handling reference points in EnergyDensityEstimator

### PR DESCRIPTION
**Pull request by PDoakORNL**
## Proposed changes

Introduces `NEReferencePoints` a necessary sub component of the ported EnergyDensityEstimator. 
A `ReferencePointsInput` class and testing for both.

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- New feature
- Build related changes

### Does this introduce a breaking change?

- Yes

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes/No. This PR is up to date with current the current state of 'develop'
- Yes/No. Code added or changed in the PR has been clang-formatted
- Yes/No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Yes/No. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by ye-luo** [comment]
> Introduces NEReferencePoints a necessary sub component of the ported EnergyDensityEstimator.

Do you mean this is a class only to be consumed by EnergyDensityEstimator?

File: src/Estimators/NEReferencePoints.h [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+#include <Configuration.h>
+#include "OhmmsData/OhmmsElementBase.h"
+#include "Particle/ParticleSet.h"
+#include "QMCHamiltonians/ObservableHelper.h"
+#include "OhmmsPETE/Tensor.h"
+#include "ReferencePointsInput.h"
+
+namespace qmcplusplus
+{
+class NEReferencePoints
```
Comment (ye-luo) [inline src/Estimators/NEReferencePoints.h:27]: I have a hard time to figure out what this class does and what it is for. Need documentation.

File: src/Estimators/NEReferencePoints.h [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+private:
+  Axes axes;
+  ReferencePointsInput input_;
+};
+
+std::ostream& operator<<(std::ostream& out, const NEReferencePoints& rhs);
+
+namespace testing
+{
+class TestableNEReferencePoints : public NEReferencePoints
```
Comment (ye-luo) [inline src/Estimators/NEReferencePoints.h:55]: If this is for testing. Can it be moved to the test source file?

File: src/Estimators/ReferencePointsInput.h [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+#include <OhmmsPETE/Tensor.h>
+#include <OhmmsPETE/TinyVector.h>
+#include "InputSection.h"
+
+namespace qmcplusplus
+{
+namespace testing
+{
+template<typename T>
+class EnergyDensityTests;
```
Comment (ye-luo) [inline src/Estimators/ReferencePointsInput.h:29]: I don't see it's been consumed in this file. Is there a better place to host these lines?

File: src/Estimators/NEReferencePoints.cpp [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+#include "NEReferencePoints.h"
+#include "Utilities/string_utils.h"
+#include "OhmmsData/AttributeSet.h"
+#include "QMCHamiltonians/ObservableHelper.h"
+
+namespace qmcplusplus
+{
+NEReferencePoints::NEReferencePoints(const ReferencePointsInput& rp_input,
+                                     ParticleSet& pset,
```
Comment (ye-luo) [inline src/Estimators/NEReferencePoints.cpp:23]: Does this class modify `ParticleSet`? if not, put const.

File: src/Estimators/tests/CMakeLists.txt [hunk 5]
```diff
@@ -37,10 +37,19 @@ set(SRCS
...
     test_MagnetizationDensityInput.cpp
-    test_MagnetizationDensity.cpp)
+    test_MagnetizationDensity.cpp
+    test_PerParticleHamiltonianLogger.cpp
+    test_ReferencePointsInput.cpp
+    test_ReferencePoints.cpp
+    )
+
+# Tests incompatible with DiracDeterminantCUDA
+# DiracDeterminantsCUDA cannot be copied
```
Comment (ye-luo) [inline src/Estimators/tests/CMakeLists.txt:49]: we don't have `DiracDeterminantsCUDA` any more.

File: src/Estimators/tests/CMakeLists.txt [hunk 6]
```diff
@@ -37,10 +37,19 @@ set(SRCS
...
+if(NOT QMC_CUDA)
```
Comment (ye-luo) [inline src/Estimators/tests/CMakeLists.txt:50]: `QMC_CUDA` is gone.

File: src/Estimators/tests/CMakeLists.txt [hunk 7]
```diff
@@ -82,7 +92,7 @@ if(HAVE_MPI)
       platform_omptarget_LA
       utilities_for_test)
   endif()
-  target_link_libraries(${UTEST_EXE} catch_main qmcestimators)
+  target_link_libraries(${UTEST_EXE} catch_main qmcestimators qmcwfs qmcparticle qmcham qmcwfs_omptarget qmcutil utilities_for_test)
```
Comment (ye-luo) [inline src/Estimators/tests/CMakeLists.txt:95]: I don't see the need of change in the above two `target_link_libraries`

File: src/Estimators/NEReferencePoints.h (hunk 1 above)
Comment (PDoakORNL) [reply src/Estimators/NEReferencePoints.h:27]: Yes but I'll give it a little more doc.

File: src/Estimators/tests/CMakeLists.txt (hunk 6 above)
Comment (PDoakORNL) [reply src/Estimators/tests/CMakeLists.txt:50]: Removed conditional sources

**Comment by PDoakORNL** [comment]
@ye-luo  thanks for the review must recent push should satisfy all your comments.  CMake changes reduced, some may return incrementally in future PR's but some were probably uneeded cruft from debugging that I'm glad you noticed.

**Comment by PDoakORNL** [comment]
test this please

**Comment by PDoakORNL** [comment]
test this please

**Comment by PDoakORNL** [comment]
Yes it is only consumed by EnergyDensityEstimator.  But someone could easily use it as part of a new estimator in concert with NESpaceGrid.  It’s pretty clear this is what Jaron was thinking with these.

**Comment by ye-luo** [comment]
Is `ReferenceLocations` a name better than `ReferencePoints`?

**Comment by PDoakORNL** [comment]
Doesn't make a difference to me, I'm using the name from the original code. I certainly don't mind doing a rename refactor at a later point.

**Comment by ye-luo** [comment]
> Doesn't make a difference to me, I'm using the name from the original code. I certainly don't mind doing a rename refactor at a later point.

Since the xml input calls it `reference_points`, just keep the current name.


What does NE stand for?

**Comment by PDoakORNL** [comment]
Abbreviation of New to prevent ODR issues with QMCHamiltonian version.

**Comment by PDoakORNL** [comment]
intention is to move them to Non prefixed names when legacy EnergyDensityOperator is removed.

**Comment by ye-luo** [comment]
> intention is to move them to Non prefixed names when legacy EnergyDensityOperator is removed.

Could you add a FIXME in the comments of these classes about this intended future change.

**Comment by PDoakORNL** [comment]
test this please

**Comment by PDoakORNL** [comment]
test this please

**Review by ye-luo — APPROVED** [review]

**Comment by PDoakORNL** [comment]
test this please

**Pull request merged**